### DISCUSSION TITLE
Function for LIdar

### DISCUSSION THREAD

**Discussion by brunodcassio** [Guilouz/Creality-Helper-Script-Wiki#527] [Features Request]
Is it possible to give a useful function to the Lidar sensor? I feel that on the K1max it is just serving as weight for the print head.

**Comment by mikegriffin84** [comment]
Doesn't the LIDAR do bed leveling and flow calculations??? I think it is not just extra weight....

**Reply by brunodcassio** [reply]
only using the Creality Print slicer, which is inconvenient as it is a bad slicer. If this function could be indexed to the Orca Slicer, it would make more sense, as it is a far superior slicer.

**Comment by Guilouz** [comment]
Lidar is only used for pressure advance calculation and first layer verification.
This is working also with gcode from OrcaSlicer.

**Reply by brunodcassio** [reply]
Sorry, but I don't know how to do it. Can you share this code and where to add it?

**Comment by Guilouz** [comment]
You just need to start print from screen and checking calibration box

**Reply by brunodcassio** [reply]
Is this function on the slicer or on the printer? On the printer I have already selected the pressure advance test, so much so that it is marked in parentheses that it only works with Creality Print. On the Orca Slicer I didn't find any button to activate. Sorry.

**Reply by brunodcassio** [reply]
I think I understand, in this case I must generate the G-Code of my printing on the Orca Slicer and have it printed via Creality Cloud for the Lidar to work. Will it be possible to perform this function in the future by sending the file directly from Orca Slicer to Mainsail or Fluid without it going through a third program?

**Reply by Tombraider2006** [reply]
i found this answer in one trad but i not have k1max to make test. 
> Hello colleagues.
> Some time ago I asked a question about why Lidar is needed, how it works and how to launch it in a controlled manner. BUT, unfortunately, I did not receive a complete answer.
> I conducted my own research and am sharing the results with you.
> 
> The lidar in K1 MAX performs two types of calibrations:
> (1) - PA (Pressure Advance) is calibrated, while several zigzag lines are printed on the left side of the table and scanned with a lidar.
> (2) - the Flow is calibrated, while the first layer of the model is printed, the result is scanned and evaluated; if it goes badly, a message is displayed on the screen and the printer stops. If it lays down normally, but there is a slight deviation in extrusion, then the flow is adjusted (the printer pauses for a while), after which printing continues.
> 
> (1) - for PA calibration to start, the following conditions must be met:
> >g-code should not contain commands for setting PA, that is, in the slicer it is necessary to REMOVE the checkbox responsible for setting PA in the rod (filament) settings block.
> In CrealityPrint this is "Enable pressure advance"
> In OrcaSlicer it is "Enable Pressure Advance"
> And also make sure that this parameter is not set in user settings blocks (for example, in the G-code of the beginning of the Filament, etc. scripts)
> >it is necessary that in g-code the following mandatory lines (parameters) are present/set in the starting block:
> ```
> ;
> ;Outer Wall Speed:[outer_wall_speed]
> ;Outer Wall Acceleration:[outer_wall_acceleration]
> ;Acceleration to decelerate:[accel_to_decel_factor]
> ;
> ```
> Here I have provided a fragment from the starting G-code for OrcaSlicer, for CrealityPrint this is not necessary (it sets it itself), for other slicers you need to substitute the variables that are responsible for these parameters.
> 
> The PA obtained by the calibration is remembered by the printer until the filament is changed (or rebooted), so the calibration may not be restarted.
> If the above conditions are not met, then at the beginning of calibration the screen will write something like: “this file does not support calibration...”
> 
> (2) - in order for the flow calibration to be performed, it is necessary and sufficient that there are no parameters in the g-code that set/change the flow (Flow), so it is enough to set flow = 100% or 1 (depending on which slicer) in the bar parameters.
> 
> Both calibrations are guaranteed to start if you start printing from the printer screen, having previously dropped a file on it, with the “calibration” checkbox checked. When you start printing from the Web interface with CrealityPrint, it usually works, but not with other slicers. Well, or you need to specifically write commands into the start g-code.

**Reply by brunodcassio** [reply]
I don't understand where to put this initial code, can you show an example within Orca Slicer?

**Reply by jlanton0000** [reply]
so, no option to start from Orca /Slicer directly. you must upload the Gcode and start printing from the screen cheking the calibrate option first, right?

**Reply by Tombraider2006** [reply]
![2024-06-11_13-14-21](https://github.com/Guilouz/Creality-Helper-Script-Wiki/assets/59514540/b30ff161-a7ee-42ab-b66f-941f0fa88f57)

**Comment by kanishk619** [comment]
I did some digging and found that you can connect to port 9999 with websocket and trigger print with AI support. Upload the file to printer and then trigger using below command.

```shell
echo '{"method":"set","params":{"opGcodeFile":"printprt:/usr/data/printer_data/gcodes/filename_to_be_printed.gcode","enableSelfTest":1,"aiPausePrint":1,"aiDetection":1,"aiFirstFloor":1,"aiSw":1}}' | ./websocat.x86_64-unknown-linux-musl ws://printer.ip.address:9999

```

You should also remove the existing mesh data by using the below command.
```shell
echo '{"method":"set","params":{"rmProbedMatrix":1}}' | ./websocat.x86_64-unknown-linux-musl ws://192.168.68.107:9999
```

Basically all the aI parameters are just to control the settings which you can also change from screen. Main parameter to trigger print calibration (as per creality print) is "enableSelfTest" set to 1. I wasn't able to find anyway to control the "Motion Advance" option (the parameter name is flowControl).

aiFirstFloor = First layer detection
aiPausePrint = Printing paused due to faults
aiDetection = AI detection option
aiSw = accept aiDetection disclaimer

Hoping this can be integrated somehow in START_PRINT gccode macro

**Reply by brunodcassio** [reply]
Honestly I don't understand how to apply this to my printer, I hope you're right and that this function will be included in the update codes in the future.

**Comment by NexGen-3D-Printing** [comment]
Just got this printer and was looking at the use case for the Lidar, from what I gleamed here, I think I will just remove it, as it is extra weight hanging off the tool head and has no practical use, the K1C doesn't need it, and the latest Max is just a big K1C.

**Reply by kanishk619** [reply]
Even I removed it
